### ISSUE TITLE
Git tries to access removed file after change in case (with setCaseSensitiveInfo)

### ISSUE THREAD

**Issue by tali** [git-for-windows/git#3686]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.34.1.windows.1
cpu: x86_64
built from commit: 2ca94ab318509b3c271e82889938816bad76dfea
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19042.1466]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option: main
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Rebase
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Enabled WSL.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash, but same behavior can be reproduced using PowerShell.

 - What commands did you run to trigger this issue? 

```
$ fsutil.exe file setCaseSensitiveInfo .
$ git init
$ echo foo > bar
$ git add bar
$ rm bar
$ git status
Changes to be committed:
        new file:   bar
Changes not staged for commit:
        deleted:    bar
$ echo foo > BAR
$ git status
Changes to be committed:
        new file:   bar
Changes not staged for commit:
        modified:    bar
Untracked files:
        BAR
$ git add -A
error: open("bar"): No such file or directory
error: unable to index file 'bar'
fatal: updating files failed
```
 - What did you expect to occur after running these commands?

After creating a new file with same name but different case, the old file should still be recognized as 'deleted' (not 'modified')
and `git add` should succeed in removing it from the index.

 - What actually happened instead?

After deleting and recreating the file with a different case, Git suddenly thinks the old (already removed) file name still exists.
Then it tries to access that file and fails.
The same happens when renaming the file outside of Git instead of deleting/recreating.

**Comment by dscho** [comment]
I started an attempt to reproduce this issue:

```console
$ fsutil.exe file setCaseSensitiveInfo .
Error:  The directory is not empty.
```

☹️

**Comment by dscho** [comment]
@tali I have a tentative fix over in #3687 and triggered a build over [here](https://github.com/git-for-windows/git/actions/runs/1808173487) which, once finished, should offer you an installer and a portable Git to test.

**Comment by dscho** [comment]
@tali please test either [the installer](https://github.com/git-for-windows/git/suites/5211252175/artifacts/159411713) or [the portable Git](https://github.com/git-for-windows/git/suites/5211252175/artifacts/159411715), to verify that my patch fixes the bug also on your side.

**Comment by tali** [comment]
This bug is fixed by #3687.
